### PR TITLE
Fix for softmax accuracy issues

### LINKED ISSUE #37841: PCC drop in ttnn.softmax PCC=0.5832070037132617
### Describe the bug
ttnn. softmax having pcc drop = 0.5832070037132617
The affected models are listed below:
```
test_all_models_torch[gemma/pytorch-google/gemma-1.1-7b-it-single_device-inference]                                                         language    red         p150         INCORRECT_RESULT           PASSING       0.9724180883623303      0.99       PCC_EN   single_device    131.686   N/A
test_all_models_torch[llama/causal_lm/pytorch-llama_3_1_8b_instruct-single_device-inference]                                                language    red         p150         INCORRECT_RESULT           PASSING       0.9702271827420873      0.99       PCC_EN   single_device    171.237   N/A
test_all_models_torch[mistral/pytorch-mistral_nemo_instruct_2407-single_device-inference]                                                   language    red         p150         INCORRECT_RESULT           PASSING       0.9821535219381753      0.99       PCC_EN   single_device    187.107   N/A
test_all_models_torch[phi2/causal_lm/pytorch-microsoft/phi-2-single_device-inference]                                                       language    red         n150         INCORRECT_RESULT           PASSING       0.9344665590421978      0.98       PCC_EN   single_device    142.625   N/A
test_all_models_torch[phi2/causal_lm/pytorch-microsoft/phi-2-single_device-inference]                                                       language    red         p150         INCORRECT_RESULT           PASSING       0.9310092454805221      0.98       PCC_EN   single_device    109.987   N/A
```

### Steps to Reproduce
```
git checkout sonalib/softmax
git submodule update --init --recursive
git lfs pull --include ‘tensors/*.tensorbin'
pytest /tt-metal/tests/ttnn/unit_tests/operations/softmax/main.py -svv
```

### Logs
[softmax_ttnn.log](https://github.com/user-attachments/files/25290039/softmax_ttnn.log)

### PR DESCRIPTION
**Pull request by fplavecTT**
### Ticket
#37841

### Problem description
For a given input tensor with extreme input values, `ttnn.softmax` PCC against Pytorch is very low (0.583)

### What's changed
The issue was root caused to the `calc_numeric_stable` function. While calls to SUM reduce operations were correctly passing the `ENABLE_FP32_DEST_ACC` flag, calls to MAX reduce were not. Added the flag to calls to MAX reduce.
With the fix, PCC goes up to 0.931 for the input provided in the issue.

### Checklist

- [x] [All post-commit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml?query=branch:fplavec_37841_softmax)
- [x] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:fplavec_37841_softmax)
  ✅ No new failures! (2 pre-existing failures on main)
  ⚠️  SAME FAILURES (pre-existing on main):
     ❌ blackhole-multi-card-fast-unit-tests (P300-viommu) / blackhole P300-viommu Host IO fast unit test (viommu only)
     ❌ ops-unit-tests (P150b) / blackhole deepseek blitz op tests blackhole P150b
- [x] [cpp-unit-tests](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:fplavec_37841_softmax)
  ✅ No new failures! (2 pre-existing failures on main)
  ⚠️  SAME FAILURES (pre-existing on main):
     ❌ cpp-unit-tests (blackhole, P150b) / tools blackhole P150b
     ❌ cpp-unit-tests (blackhole, P100a) / tools blackhole P100a
- [x] New/Existing tests provide coverage for changes
  Verified that unit test provided in the issue achieves PCC of 0.931 with the fix

#### Model tests

If your changes cover model-related code, you should run tests corresponding to affected models and platforms (Single card, T3K, Galaxy). "Choose your pipeline" workflows facilitate running multiple kinds of tests in a single run. Each offers `models-mandatory` and `models-extended` presets.
The former includes a minimal set of tests, to be run always. The latter extends that with additional ones - use your best judgement in deciding which is the most appropriate for your PR.

- [ ] [(Single) Choose your pipeline](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:fplavec_37841_softmax)
 There are some new failures here, though they look like infra issues. I'll try rerunning to confirm.
  - [x] `models-mandatory` preset (runs: [Device perf regressions](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) and [Frequent model and ttnn tests](https://github.com/tenstorrent/tt-metal/actions/workflows/fast-dispatch-full-regressions-and-models.yaml))
  - [x] `models-extended` preset (runs: the mandatory tests, plus [Demo](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) and [Model perf](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) tests)
  - [ ] other selection - specify runs

- [ ] [(T3K) Choose your pipeline](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:fplavec_37841_softmax)
  - [ ] `models-mandatory` preset (runs: [Unit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml))
  - [ ] `models-extended` preset (runs: the mandatory tests, plus [Demo](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml) and [Model perf](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-model-perf-tests.yaml) tests)
  - [ ] other selection - specify runs

- [ ] [(Galaxy) Choose your pipeline](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:fplavec_37841_softmax)
  - [ ] `models-mandatory` preset (runs: [Quick tests](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-quick.yaml))
  - [ ] `models-extended` preset (runs: the mandatory tests, plus [Demo](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-demo-tests.yaml) and [Model perf](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-perf-tests.yaml) tests)
  - [ ] other selection - specify runs

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/normalization/softmax/device/kernels/attention/compute/softmax.cpp [hunk 1]
```diff
@@ -27,11 +27,12 @@ void calc_numeric_stable(
     reconfig_data_format(cb_in, cb_bcast_scaler);
     cb_reserve_back(cb_max, 1);
     cb_wait_front(cb_bcast_scaler, 1);
-    reduce_init<PoolType::MAX, ReduceDim::REDUCE_ROW>(cb_in, cb_bcast_scaler, cb_max);
+    reduce_init<PoolType::MAX, ReduceDim::REDUCE_ROW, ENABLE_FP32_DEST_ACC>(cb_in, cb_bcast_scaler, cb_max);
     for (uint32_t wt = 0; wt < Wt; wt++) {
         cb_wait_front(cb_in, wt + 1);
         constexpr uint32_t bcast_scaler0 = 0;
-        reduce_tile<PoolType::MAX, ReduceDim::REDUCE_ROW>(cb_in, cb_bcast_scaler, wt, bcast_scaler0, 0);
+        reduce_tile<PoolType::MAX, ReduceDim::REDUCE_ROW, ENABLE_FP32_DEST_ACC>(
```
Comment (bbradelTT) [inline ttnn/cpp/ttnn/operations/normalization/softmax/device/kernels/attention/compute/softmax.cpp:34]: Please also update reduce_uninit.

File: ttnn/cpp/ttnn/operations/normalization/softmax/device/kernels/attention/compute/softmax_sharded.cpp [hunk 2]
```diff
@@ -20,11 +20,12 @@ ALWI void calc_numeric_stable(uint32_t cb_in, uint32_t cb_bcast_scaler, uint32_t
...
-    reduce_init<PoolType::MAX, ReduceDim::REDUCE_ROW>(cb_in, cb_bcast_scaler, cb_max);
+    reduce_init<PoolType::MAX, ReduceDim::REDUCE_ROW, ENABLE_FP32_DEST_ACC>(cb_in, cb_bcast_scaler, cb_max);
     cb_wait_front(cb_bcast_scaler, 1);
     for (uint32_t w = 0; w < block_w; w++) {
         constexpr uint32_t bcast_scaler0 = 0;
-        reduce_tile<PoolType::MAX, ReduceDim::REDUCE_ROW>(cb_in, cb_bcast_scaler, w, bcast_scaler0, 0);
+        reduce_tile<PoolType::MAX, ReduceDim::REDUCE_ROW, ENABLE_FP32_DEST_ACC>(
+            cb_in, cb_bcast_scaler, w, bcast_scaler0, 0);
     }
     reduce_uninit();
```
[the remark this answers is not in the document]
Comment (fplavecTT) [reply ttnn/cpp/ttnn/operations/normalization/softmax/device/kernels/attention/compute/softmax_sharded.cpp]: Done.

File: ttnn/cpp/ttnn/operations/normalization/softmax/device/kernels/attention/compute/softmax.cpp (hunk 1 above)
Comment (fplavecTT) [reply ttnn/cpp/ttnn/operations/normalization/softmax/device/kernels/attention/compute/softmax.cpp:34]: Updated with both template argument and CB argument. Also updated the `reduce_uninit` for the SUM further below, which was also missing both.

File: ttnn/cpp/ttnn/operations/normalization/softmax/device/kernels/attention/compute/softmax.cpp [hunk 3]
```diff
@@ -27,11 +27,12 @@ void calc_numeric_stable(
...
+            cb_in, cb_bcast_scaler, wt, bcast_scaler0, 0);
     }
     reduce_uninit();
```
[the remark this answers is not in the document]
Comment (fplavecTT) [reply ttnn/cpp/ttnn/operations/normalization/softmax/device/kernels/attention/compute/softmax.cpp]: Done

**Review by bbradelTT — APPROVED** [review]
